File: common/src/main/java/org/valkyrienskies/clockwork/content/contraptions/resistor/RedstoneResistorBlockEntity.java
package org.valkyrienskies.clockwork.content.contraptions.resistor;

import com.simibubi.create.content.contraptions.RotationPropagator;
import com.simibubi.create.content.contraptions.base.KineticTileEntity;
import com.simibubi.create.content.contraptions.goggles.IHaveGoggleInformation;
import com.simibubi.create.content.contraptions.relays.advanced.sequencer.SequencedGearshiftBlock;
import com.simibubi.create.content.contraptions.relays.encased.SplitShaftTileEntity;
import com.simibubi.create.foundation.utility.Iterate;
import com.simibubi.create.foundation.utility.Lang;
import com.simibubi.create.foundation.utility.animation.LerpedFloat;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.ticks.TickPriority;

import java.util.List;

public class RedstoneResistorBlockEntity extends SplitShaftTileEntity implements IHaveGoggleInformation {

    int redstoneLevel = 0;

    public RedstoneResistorBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state) {
        super(type, pos, state);
    }

    public int getRedstoneLevel() {
        return redstoneLevel;
    }

    public void onRedstoneUpdate(int redstoneLevel) {
        if (redstoneLevel == this.redstoneLevel) return;

        this.redstoneLevel = redstoneLevel;

        this.setChanged();
        this.detachKinetics();
    }

    @Override
    public void write(CompoundTag compound, boolean clientPacket) {
        compound.putInt("RedstoneLevel", redstoneLevel);
        super.write(compound, clientPacket);
    }


    @Override
    protected void read(CompoundTag compound, boolean clientPacket) {
        redstoneLevel = compound.getInt("RedstoneLevel");
        super.read(compound, clientPacket);
    }

    @Override
    public float getRotationSpeedModifier(Direction face) {
        if (isVirtual())
            return 1;
        if (!hasSource() || face == getSourceFacing())
            return 1;

        return (Math.abs(redstoneLevel-15))/15f;
    }

    @Override
    public boolean addToGoggleTooltip(List<Component> tooltip, boolean isPlayerSneaking) {
        tooltip.add(componentSpacing.plainCopy().append(Lang.translateDirect("tooltip.analogStrength", this.redstoneLevel)));

        return true;
    }

    public int getState() {
        return redstoneLevel;
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/kinetics/resistor/RedstoneResistorBlockEntity.kt
package org.valkyrienskies.clockwork.content.kinetics.resistor

import com.simibubi.create.api.equipment.goggles.IHaveGoggleInformation
import com.simibubi.create.content.kinetics.RotationPropagator
import com.simibubi.create.content.kinetics.transmission.SplitShaftBlockEntity
import com.simibubi.create.foundation.utility.CreateLang
import net.createmod.catnip.data.Iterate
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Component
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.ticks.TickPriority
import org.valkyrienskies.clockwork.util.ClockworkConstants
import kotlin.math.abs

open class RedstoneResistorBlockEntity(type: BlockEntityType<*>?, pos: BlockPos, state: BlockState) :
    SplitShaftBlockEntity(type, pos, state), IHaveGoggleInformation {
    var state = 0
    var lastChange = 0
    override fun tick() {
        super.tick()
        lastChange = state
        state = getPower(level!!, worldPosition)
        if (state != lastChange) {
            detachKinetics()
        }
    }

    override fun detachKinetics() {
        RotationPropagator.handleRemoved(level, worldPosition, this)

        // Re-attach next tick
        level!!.scheduleTick(worldPosition, blockState.block, 0, TickPriority.EXTREMELY_HIGH)
    }

    public override fun write(compound: CompoundTag, clientPacket: Boolean) {
        compound.putInt(ClockworkConstants.Nbt.REDSTONE_LEVEL, state)
        compound.putInt(ClockworkConstants.Nbt.CHANGE_TIMER, lastChange)
        super.write(compound, clientPacket)
    }

    override fun read(compound: CompoundTag, clientPacket: Boolean) {
        state = compound.getInt(ClockworkConstants.Nbt.REDSTONE_LEVEL)
        lastChange = compound.getInt(ClockworkConstants.Nbt.CHANGE_TIMER)
        super.read(compound, clientPacket)
    }

    private fun getPower(worldIn: Level, pos: BlockPos): Int {
        var power = 0
        for (direction in Iterate.directions) power =
            Math.max(worldIn.getSignal(pos.relative(direction), direction), power)
        for (direction in Iterate.directions) power =
            Math.max(worldIn.getSignal(pos.relative(direction), Direction.UP), power)
        return power
    }

    override fun getRotationSpeedModifier(face: Direction): Float {
        if (hasSource()) {
            if (face != sourceFacing) {
                val i = abs(state - 15) / 15f;
                return i
            }
        }
        return 1f
    }

    override fun addToGoggleTooltip(tooltip: MutableList<Component>, isPlayerSneaking: Boolean): Boolean {
        tooltip.add(
            CreateLang.translateDirect(
                "tooltip.analogStrength",
                state
            )
        )
        return true
    }
}
